modules/payment/tests/src/Unit/CreditCardTest.php

```diff
@@ -0,0 +1,112 @@
+<?php
+
+namespace Drupal\Tests\commerce_payment\Unit;
+
+use Drupal\commerce_payment\CreditCard;
+use Drupal\commerce_payment\CreditCardType;
+
+/**
+ * Unit test for CreditCard.
+ *
+ * @coversDefaultClass \Drupal\commerce_payment\CreditCard
+ */
+class CreditCardTest extends \PHPUnit_Framework_TestCase {
+
+  /**
+   * @covers ::getTypes
+   */
+  public function testGetTypes() {
+    $types = CreditCard::getTypes();
+    $this->assertInternalType('array', $types);
+    foreach ($types as $key => $cardType) {
+      $this->assertInstanceOf(CreditCardType::class, $cardType);
+      $this->assertEquals($key, $cardType->getId());
+    }
+  }
+
+  /**
+   * @covers ::getType
+   */
+  public function testGetInvalidType() {
+    $this->setExpectedException(\InvalidArgumentException::class, 'Invalid credit card type "monkey"');
+    CreditCard::getType("monkey");
+  }
+
+  /**
+   * @covers ::getType
+   */
+  public function testGetType() {
+    $visa = CreditCard::getType('visa');
+    $this->assertInstanceOf(CreditCardType::class, $visa);
+    $this->assertEquals('visa', $visa->getId());
+  }
+
+  /**
+   * @covers ::detectType
+   * @covers ::matchPrefix
+   * @dataProvider cardsProvider
+   */
+  public function testDetectType($number, $cardType) {
+    // Card number has to be an integer.
+    $card = CreditCard::detectType('4111 1111 1111 1111');
```

Comment:
Feels odd to test the non-integer number for every $number / $type combination.
I suggest passing a $valid variable from the data provider, then extending the data provider to pass a few invalid values (non-integer, number too long, right prefix but wrong number, unrecognized type).